tools/integration-test/src/tests/client_upgrade.rs

```diff
@@ -0,0 +1,577 @@
+            .and_then(|deposit_params| deposit_params.as_object_mut());
+
+        if let Some(max_deposit_period) = max_deposit_periods {
+            max_deposit_period.insert(
+                "max_deposit_period".to_owned(),
+                Value::String(MAX_DEPOSIT_PERIOD.to_string()),
+            );
+        } else {
+            return Err(Error::generic(eyre!(
+                "failed to update max_deposit_period in genesis file"
+            )));
+        }
+
+        let voting_period = genesis
+            .get_mut("app_state")
+            .and_then(|app_state| app_state.get_mut("gov"))
+            .and_then(|gov| gov.get_mut("voting_params"))
+            .and_then(|voting_params| voting_params.as_object_mut());
+
+        if let Some(voting_period) = voting_period {
+            voting_period.insert(
+                "voting_period".to_owned(),
+                Value::String(VOTING_PERIOD.to_string()),
+            );
+            Ok(())
+        } else {
+            Err(Error::generic(eyre!(
+                "failed to update voting_period in genesis file"
+            )))
+        }
+    }
+}
+
+impl BinaryChainTest for HeightTooHighClientUpgradeTest {
+    fn run<
+        ChainA: ibc_test_framework::prelude::ChainHandle,
+        ChainB: ibc_test_framework::prelude::ChainHandle,
+    >(
+        &self,
+        _config: &ibc_test_framework::prelude::TestConfig,
+        _relayer: ibc_test_framework::prelude::RelayerDriver,
+        chains: ibc_test_framework::prelude::ConnectedChains<ChainA, ChainB>,
+    ) -> Result<(), ibc_test_framework::prelude::Error> {
+        let foreign_clients = chains.clone().foreign_clients;
+
+        let src_chain_config = chains.handle_b().config().unwrap();
+        let dst_chain_config = chains.handle_a().config().unwrap();
+
+        let upgraded_chain_id = chains.chain_id_a().0.clone();
+
+        let src_client_id = foreign_clients.client_id_b().0.clone();
+
+        // Create and send an chain upgrade proposal
+        let opts = UpgradePlanOptions {
+            dst_chain_config,
+            src_chain_config,
+            src_client_id,
+            amount: 10000000u64,
+            denom: "stake".to_string(),
+            height_offset: DELTA_HEIGHT,
+            upgraded_chain_id,
+            upgraded_unbonding_period: None,
+            upgrade_plan_name: "plan".to_string(),
+        };
+
+        build_and_send_ibc_upgrade_proposal(
+            chains.handle_a().clone(),
+            chains.handle_b().clone(),
+            &opts,
+        )
+        .map_err(Error::upgrade_chain)?;
+
+        // Wait for the proposal to be processed
+        std::thread::sleep(Duration::from_secs(2));
+
+        let driver = chains.node_a.chain_driver().0;
+
+        // Retrieve the height which should be used to upgrade the client
+        let upgrade_height = driver
+            .runtime
+            .block_on(query_upgrade_proposal_height(
```

Comment:
We can define extension methods in `tools/test-framework/src/chain/ext` so that here we can call `driver.query_upgrade_proposal_height()` while preserving the type tags.

Reply:
This has been addressed in this commit https://github.com/informalsystems/hermes/pull/2795/commits/b5b046afc91fc53c3d10a40115cdd0140d7b0010